ion/src/device/led.cpp

```diff
@@ -81,12 +103,53 @@ void initTimer() {
 }
 
 void shutdownTimer() {
-  TIM3.CCMR()->setOC2M(TIM::CCMR::OCM::ForceInactive);
-  TIM3.CCMR()->setOC3M(TIM::CCMR::OCM::ForceInactive);
-  TIM3.CCMR()->setOC4M(TIM::CCMR::OCM::ForceInactive);
+  setColorStatus(Ion::LED::Device::Color::RED, false);
+  setColorStatus(Ion::LED::Device::Color::GREEN, false);
+  setColorStatus(Ion::LED::Device::Color::BLUE, false);
 }
 
-void enforceState(bool red, bool green, bool blue) {
+/* Pulse width modulation mode allows you to generate a signal with a
+ * frequency determined by the value of the TIMx_ARR register and a duty cycle
+ * determined by the value of the TIMx_CCRx register. */
+
+void setPeriodAndDutyCycles(Mode mode, float dutyCycleRed, float dutyCycleGreen, float dutyCycleBlue, float period) {
+  constexpr int TIM3_FREQ = 4*1000;
+  switch (mode) {
+    case Mode::PWM:
+      /* Let's set the prescaler to 1. Increasing the prescaler would slow down the
+       * modulation, which can be useful when debugging. */
+      TIM3.PSC()->set(1);
+      TIM3.ARR()->set(PWMPeriod);
+      period = PWMPeriod;
+      break;
+    case Mode::BLINK:
+      /* We still want to do PWM, but at a rate slow enough to blink. */
+      TIM3.PSC()->set(Ion::Device::SYSBUS_FREQ / TIM3_FREQ);
+      TIM3.ARR()->set(period * TIM3_FREQ);
+      period *= TIM3_FREQ; // as we pre-scaled, we update the period
+      break;
+  }
+
+  TIM3.CCR2()->set(dutyCycleRed*period);
+  TIM3.CCR3()->set(dutyCycleBlue*period);
+  TIM3.CCR4()->set(dutyCycleGreen*period);
+}
+
+void setColorStatus(Color color, bool enable) {
```

Comment:
I quite don't like this function. It yields a function call and a switch when previously we only required a simple write at a given address.